subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/JavaLanguageVersion.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.jvm.toolchain;
+
+import org.gradle.api.Incubating;
+import org.gradle.jvm.toolchain.internal.DefaultJavaLanguageVersion;
+
+/**
+ * Represents a Java Language version
+ *
+ * @since 6.7
+ */
+@Incubating
+public interface JavaLanguageVersion extends Comparable<JavaLanguageVersion> {
```

Comment:
As discussed before, I'm on the fence of introducing this here. For me, it has no clear value over `JavaVersion` (only uses major) if we don't follow through with it. Right now, we just add yet another `int` wrapper to the mix of the existing types without streamlining any of the existing places. As build author, this means I have multiple APIs, all accepting different APIs. Additionally, this leads to imperative code (`JavaLanguageVersion.of(14)`) instead of declarative (JavaVersion.VERSION_14). I wonder if it would be better to just use an `int` for the sake of DSL readability and consider streamlining java versions on a larger scale than using it in a single API. Just some thoughts...

Reply:
The idea is that the DSL support will be introduced through improvements to `Property`, in a cross cutting fashion.
`JavaVersion` and its enum aspect is too limiting IMO.
As for usage in other APIs, we cannot change existing ones. Only deprecate and replace them.